### PR TITLE
BUG/API: properly retrieve numpy's printoptions dict in utils.masked

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
This is a (probably) non-backportable follow up to #16755. See that PR for full details, but in summary:
we rely on private APIs to retrive print options from numpy, which is both fragile (the exact import statement needed has changed at least twice so far), and arguably incorrect: as I hinted to in https://github.com/astropy/astropy/pull/16755#issuecomment-2245262789, using the private API `np.get_printoptions` actually gives different results (because with private APIs it is possible to get an uninitialized -yet deterministic- value for the `legacy` option).

I propose to simplify this code to bring more stability in these area, at the cost of a minor change in how maksed arrays are represented, which I think we can get away with if we ship it in astropy 7.0.0

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Still one minor portability issue. I'll revisit this once #16755 is merged for to reduce my juggling overhead 🤹🏻‍♂️

**Comment by neutrinoceros** [comment]
This should be stable now.
Keeping this portable with numpy 1 turned out to be harder than I thought (not a huge deal, but one test in particular turned out more subtle to adjust) so I adjusted my initial goal to be 100% backward compatible on numpy 1 and only accept the slight behavior changes on numpy 2.

**Comment by neutrinoceros** [comment]
Actually, I'll take this back to draft. I'd like to take a closer look into numpy so I at least understand where exactly does the different behavior comes from.

**Comment by neutrinoceros** [comment]
After a more careful inspection of numpy's source, I observe that they never go through their own public API to get print options. My guess is that it is precisely because of how the `legacy` option is implemented: it's really an integer (i.e., `113` mimics numpy 1.23's behavior), and the default value (no legacy printing) should compare greater to any possible value *and still be future-proof*, which is why they use `sys.maxsize`, but to make its meaning clearer to users, this default value is replaced with `False` by `np.get_printoptions`. Unfortunately `False` cannot be used as the default value because it compares *lower than* any possible value users might set, which is the opposite of what we want.
In short, I think we *have* to keep using private API to respect numpy's default behavior. This is unfortunate but not unheard of in the realm of numpy extensions. I'll close this now.

**Comment by mhvk** [comment]
Thanks for checking this all! I do think there is the additional benefit that, effectively, we're just copying a piece of numpy code and making the most minimal changes to it to make it do what we need. Anyway, for now we don't have to worry about it any more...
